backend/api/src/model/block.rs

```diff
@@ -0,0 +1,176 @@
+    fn shared(&self) -> &SharedData;
+
+    fn id(&self) -> Id {
+        self.shared().id
+    }
+    fn realm_id(&self) -> Id {
+        self.shared().realm_id
+    }
+    fn index(&self) -> i32 {
+        self.shared().index
+    }
+    fn title(&self) -> Option<&str> {
+        self.shared().title.as_deref()
+    }
+}
+
+#[derive(Debug, Clone, Copy, FromSql)]
+#[postgres(name = "block_type")]
+pub(crate) enum BlockType {
+    #[postgres(name = "text")]
+    Text,
+    #[postgres(name = "videolist")]
+    VideoList,
+}
+
+#[derive(Debug, Clone, Copy, FromSql, GraphQLEnum)]
+#[postgres(name = "video_list_layout")]
+pub(crate) enum VideoListLayout {
+    #[postgres(name = "horizontal")]
+    Horizontal,
+    #[postgres(name = "vertical")]
+    Vertical,
+    #[postgres(name = "grid")]
+    Grid,
+}
+
+#[derive(Debug, Clone, Copy, FromSql, GraphQLEnum)]
+#[postgres(name = "video_list_order")]
+pub(crate) enum VideoListOrder {
+    #[postgres(name = "new_to_old")]
+    NewToOld,
+    #[postgres(name = "old_to_new")]
+    OldToNew,
+}
+
+/// Data shared by all blocks.
+#[derive(GraphQLObject)]
+pub(crate) struct SharedData {
+    pub(crate) id: Id,
+    pub(crate) realm_id: Id,
+    pub(crate) index: i32,
+    pub(crate) title: Option<String>,
+}
+
+/// A block just showing some text.
+#[derive(GraphQLObject)]
+#[graphql(impl = BlockValue)]
+pub(crate) struct Text {
+    #[graphql(skip)]
+    pub(crate) shared: SharedData,
+    pub(crate) content: String,
+}
+
+#[graphql_interface]
+impl Block for Text {
+    fn shared(&self) -> &SharedData {
+        &self.shared
+    }
+}
+
+/// A block just showing some text.
+#[derive(GraphQLObject)]
+#[graphql(impl = BlockValue)]
+pub(crate) struct VideoList {
+    #[graphql(skip)]
+    pub(crate) shared: SharedData,
+    pub(crate) series: Id,
+    pub(crate) layout: VideoListLayout,
+    pub(crate) order: VideoListOrder,
+}
+
+#[graphql_interface]
+impl Block for VideoList {
+    fn shared(&self) -> &SharedData {
+        &self.shared
+    }
+}
+
+impl BlockValue {
+    /// Fetches all blocks for the given realm from the database.
+    pub(crate) async fn fetch_for_realm(realm_key: Key, context: &Context) -> FieldResult<Vec<Self>> {
+        context.db.get()
+            .await?
+            .query_raw(
+                "select id, realm_id, type, index, title, text_content, videolist_series,
+                    videolist_layout, videolist_order
+                    from blocks
+                    where realm_id=$1
+                    order by index asc",
+                [realm_key as i64].iter().map(|x| x as &dyn ToSql),
+            )
+            .await?
+            .map_err(anyhow::Error::from)
+            .and_then(|row| async move {
+                let ty: BlockType = row.get(2);
+                let shared = SharedData {
+                    id: Id::block(row.get_key(0)),
+                    realm_id: Id::realm(row.get_key(1)),
+                    index: row.get::<_, i16>(3).into(),
```

Comment:
Why `i16` here, and `i32` in the declaration?

Reply:
We use smallint in the database. I mean sure, it's arbitrary that I said there are at most 2^16 blocks per realm. But I think it's enough :P

Reply:
Yeah, sure, makes sense.